backend/src/prereq_loader.ts

```diff
@@ -0,0 +1,250 @@
+  termObj: ScheduleTerm,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<ScheduleTerm> {
+  // the new classes.
+  const newClasses: Promise<ScheduleCourse[]> = Promise.all(
+    termObj.classes.map((course: ScheduleCourse) =>
+      prereqifyScheduleCourse(course, termObj.termId, loader)
+    )
+  );
+
+  return {
+    season: termObj.season,
+    year: termObj.year,
+    termId: termObj.termId,
+    id: termObj.id,
+    classes: await newClasses,
+    status: termObj.status,
+  };
+}
+
+/**
+ * Asynchronously adds prereqs to a provided course.
+ * @param course the course to add prereqs to
+ * @param loader the loader
+ */
+async function prereqifyScheduleCourse(
+  course: ScheduleCourse,
+  termId: number,
+  loader: DataLoader<SimpleCourse, PrereqQueryResult>
+): Promise<ScheduleCourse> {
+  // the base prereqified object
+  const prereqified: ScheduleCourse = {
+    classId: course.classId,
+    subject: course.subject,
+    numCreditsMin: course.numCreditsMin,
+    numCreditsMax: course.numCreditsMax,
+  };
+
+  let queryResult: PrereqQueryResult;
+  try {
+    // produces the class.
+    queryResult = await loader.load({
+      subject: course.subject,
+      classId: Number(course.classId),
+      termId: termId,
+    });
+  } catch (err) {
+    throw err;
+    // if we error, then return the previous course.
+    return course;
+  }
+
+  // optionally add prereqs, coreqs to object.
+  if (queryResult) {
+    queryResult.coreqs ? (prereqified.coreqs = queryResult.coreqs) : undefined;
+    queryResult.prereqs
+      ? (prereqified.prereqs = queryResult.prereqs)
+      : undefined;
+    // prereqified.numCreditsMax = queryResult.maxCredits;
+    // prereqified.numCreditsMin = queryResult.minCredits;
+    return prereqified;
+  } else {
+    return course;
+  }
+}
+
+/**
+ * Queries SearchNEU using GraphQL to look up the prereqs for each of the provided courses.
+ * @param courses the courses to lookup prereqs for
+ */
+async function queryCoursePrereqData(
+  courses: SimpleCourse[]
+): Promise<PrereqQueryResult[]> {
+  // for each one of the courses, map to a string.
+  const courseSchema: string[] = courses.map((course: SimpleCourse) => {
+    return `class(classId: ${course.classId}, subject: "${course.subject}") { 
+      occurrence(termId: ${202010}) {
+        prereqs 
+        coreqs
+        name
+      }
+    }`;
+  });
+
+  // build the query schema
+  const querySchema: string = `
+  query {
+    ${courseSchema.reduce(
+      (accumulator: string, currentValue: string, index: number) => {
+        return accumulator + `course${String(index)}: ${currentValue}\n`;
+      },
+      ""
+    )}
+  }
+  `;
+
+  // make the request.
+  let queryResult = await request({
+    uri: "https://searchneu.com/graphql",
+    method: "POST",
+    body: JSON.stringify({ query: querySchema }),
+    headers: {
+      "Content-type": "application/json",
+    },
+  });
+  totalNumRequests += 1;
```

Comment:
Same as above. Out of curiosity, how many requests does it end up making? It it called once for the length of the `SimpleCourse[]` or fewer times?

Reply:
It makes around 20 per unique plan! For BSCS, made 20 requests for the first plan, and then zero for the second and third.